drivers/gnss/gnss_u_blox_m10.c

```diff
@@ -0,0 +1,1058 @@
+/*
+ * Copyright 2024 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/gnss.h>
+#include <zephyr/drivers/gnss/gnss_publish.h>
+#include <zephyr/modem/chat.h>
+#include <zephyr/modem/ubx.h>
+#include <zephyr/modem/backend/uart.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/gpio.h>
+#include <string.h>
+
+#include "gnss_nmea0183.h"
+#include "gnss_nmea0183_match.h"
+#include "gnss_parse.h"
+
+#include "gnss_u_blox_protocol/gnss_u_blox_protocol.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ubx_m10, CONFIG_GNSS_LOG_LEVEL);
+
+#define DT_DRV_COMPAT			u_blox_m10
+
+#define UART_RECV_BUF_SZ		128
+#define UART_TRNF_BUF_SZ		128
+
+#define CHAT_RECV_BUF_SZ		256
+#define CHAT_ARGV_SZ			32
+
+#define UBX_RECV_BUF_SZ			UBX_FRM_SZ_MAX
+#define UBX_TRNS_BUF_SZ			UBX_FRM_SZ_MAX
+#define UBX_WORK_BUF_SZ			UBX_FRM_SZ_MAX
+
+#define UBX_FRM_BUF_SZ			UBX_FRM_SZ_MAX
+
+#define MODEM_UBX_SCRIPT_TIMEOUT_MS	1000
+#define UBX_M10_SCRIPT_RETRY_DEFAULT	10
+
+#define UBX_M10_GNSS_SYS_CNT		8
+#define UBX_M10_GNSS_SUPP_SYS_CNT	6
+/* The datasheet of the device doesn't specify boot time. But 1 sec helped significantly. */
+#define UBX_M10_BOOT_TIME_MS		1000
+
+enum MODEM_MODULE {
```

Comment:
```suggestion
enum modem_module {
```

if you really want an enum here, though to me it looks like you could just replace this with a boolean and everything would be come a tiny bit simpler, do you forsee this getting more values in the future? The whole `ubx_m10_modem_module_switch` just looks really really weird right now

Reply:
@fabiobaltieri We don't foresee more modem modules being used in this driver. We could utilize a boolean macro, but i thought enum improves readability. I would prefer an enum here, but if you want i will change this as you said.

Reply:
Ok, I think it should be changed, I mean this would maybe make sense if you wanted to do stuff like `ubx_m10_modem_module_switch(MODEM_MODULE_CHAT, MODEM_MODULE_CHAT);`, it's just odd to have an enum for what is really a boolean. @bjarki-trackunit?

Reply:
Yes @fabiobaltieri, I understand your point now. Passing two arguments to `ubx_m10_modem_module_switch` isn't required. Would the following work? 

```
/**
 * @brief Changes modem module (chat or ubx) attached to the uart pipe.
 * @param dev Dev instance
 * @param change_from_to 0 for changing from "chat" to "ubx", 1 for changing from "ubx" to "chat"
 * @returns 0 if successful
 * @returns negative errno code if failure
 */
static int ubx_m10_modem_module_change(const struct device *dev, bool change_from_to)
{
	struct ubx_m10_data *data = dev->data;
	int ret;

	if (change_from_to == 0) {
		modem_chat_release(&data->chat);
		ret = modem_ubx_attach(&data->ubx, data->uart_pipe);
	} else { /* change_from_to == 1 */
		modem_ubx_release(&data->ubx);
		ret = modem_chat_attach(&data->chat, data->uart_pipe);
	}

	if (ret < 0) {
		(void)modem_pipe_close(data->uart_pipe);
	}

	return ret;
}
```

Ps. I would prefer an enum here (or a couple of macros). Because without these, while calling `ubx_m10_modem_module_change`, we have to pass a literal. For now I have pushed the changes as per your comments.
```
enum modem_module_change {
	MODEM_MODULE_CHAT_TO_UBX = 0,
	MODEM_MODULE_UBX_TO_CHAT = 1,
};
```

Reply:
the function effectively attaches either chat or ubx pipe, modem_pipe_attach will automatically release the currently attached pipe, so simply attaching the required pipe should be clear and concise, without requiring a helper function at all :)